### PR TITLE
Implement exponential moving average estimator

### PR DESCRIPTION
**Pull request by awemany**
This is based on top of #1507 and extracts the moving average calculation into a new class to do it the right way(TM).

Also adds a simple unit test.

As per discussion with @dgenr8.

### CODE REVIEW

**Comment by sickpig** [comment]
Neat idea. Thanks @awemany and @dgenr8.

It seems like that we are failing while using clang to produce the binaries, see: https://travis-ci.org/BitcoinUnlimited/BitcoinUnlimited/jobs/463823492#L2267

```
../../src/test/stat_tests.cpp:151:38: error: too many arguments provided to
      function-like macro invocation
BOOST_AUTO_TEST_CASE(stat_expmovavg, *boost::unit_test::tolerance(1e-4))
                                     ^
/usr/include/boost/test/unit_test_suite.hpp:114:9: note: macro
      'BOOST_AUTO_TEST_CASE' defined here
#define BOOST_AUTO_TEST_CASE( test_name )                               \
        ^
../../src/test/stat_tests.cpp:151:1: error: C++ requires a type specifier for
      all declarations
BOOST_AUTO_TEST_CASE(stat_expmovavg, *boost::unit_test::tolerance(1e-4))
^
../../src/test/stat_tests.cpp:153:5: error: 'TestExponentialMovingAverage' does
      not refer to a value
    TestExponentialMovingAverage ema;
    ^
../../src/test/stat_tests.cpp:136:7: note: declared here
class TestExponentialMovingAverage : public ExponentialMovingAverage
      ^
../../src/test/stat_tests.cpp:176:2: error: expected ';' after top level
      declarator
}
 ^
 ;
```

**Comment by awemany** [comment]
@sickpig: Thanks for digging that out. It looks like this is due to me using too-new `libboost` features. I have added a commit that should make it compatible with `libboost1.58` again. Let's see how it goes.

**Review by sickpig — APPROVED** [review]

**Comment by awemany** [comment]
As a further test, I just confirmed its approximately correct behavior in the mempool TXN estimation, using

`./bitcoin-cli -regtest generate 1000`

and then `for i in `\``seq 0 100`\``; do ./bitcoin-cli -regtest sendtoaddress <bch-addr> 0.0001; ./bitcoin-cli -regtest getmempoolinfo; done`

building the txn rate up, followed by `watch ./bitcoin-cli -regtest getmempoolinfo` and watching the txn rate decay again.

File: src/stat.cpp [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#include <cmath>
+#include "utiltime.h"
+#include "stat.h"
+
+
+ExponentialMovingAverage::ExponentialMovingAverage(const double _interval,
+    const size_t _time_units) :
+    interval(_interval), time_units(_time_units) {}
```
Comment (dagurval) [inline src/stat.cpp:8]: suggest throwing `std::invalid_argument` if `_interval == 0 || _time_units == 0`. You risk divide by zero in `::update` if not.

File: src/stat.cpp [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+int64_t ExponentialMovingAverage::time() {
+    return GetTimeMicros();
+}
+
+void ExponentialMovingAverage::update(const size_t counts) {
+    std::lock_guard<std::mutex> lock(cs);
```
Comment (dagurval) [inline src/stat.cpp:15]: Nit: You could take the lock _after_ calling time `time()`

File: src/test/stat_tests.cpp [hunk 3]
```diff
@@ -131,4 +132,47 @@ BOOST_AUTO_TEST_CASE(stat_empty_construct)
...
+    TestExponentialMovingAverage()
+        : ExponentialMovingAverage(3.0, // 3s decay time
+              1000 // unit 'milliseconds'
+              )
+    {
+    }
+    void set_time(const int64_t _t) { t = _t; }
+protected:
+    int64_t t = 0;
+    int64_t time() { return t; }
```
Comment (dagurval) [inline src/test/stat_tests.cpp:148]: Nit: Add `override`

File: src/stat.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    int64_t delta = now - last_update;
+    if (delta > 0) // guard against clock resets
+        rate *= exp(- (delta / (time_units * interval)));
+    last_update = now;
+    rate += counts / interval;
+}
+
+double ExponentialMovingAverage::value() {
+    update(0); // decay rate to current value
```
Comment (dagurval) [inline src/stat.cpp:28]: A side effect in a getter function 🤔

File: src/stat.h [hunk 5]
```diff
@@ -681,4 +681,47 @@ class LinearHistogram
...
+    double interval;
+
+    //! Time units as set by constructor
+    int64_t time_units;
+
+protected:
+    //! Function used for time calculations. Can be overwritten e.g. for unit testing
+    virtual int64_t time();
+    //! Lock for serializing updates
+    std::mutex cs;
```
Comment (dagurval) [inline src/stat.h:723]: Nit: Make `cs` private

**Pull request closed without merging**